Add "no-new-attributes" behavior to our models

We should add "no-new-attributes" behavior to all our models in `pip._internal.models`. It'd be easy to enforce this constraint on them using the interpreter than manually in code reviews. :)

Two of the ways to achieve this using just the standard library are:

- `__slots__`
- `collections.namedtuple`
- (both together?)

In case someone (eg. a future me) wants to look up how __slots__ work, take a look at https://stackoverflow.com/a/28059785/1931274.

_Originally posted by @pradyunsg in https://github.com/pypa/pip/pull/7310_

The `attrs` module might be too dynamic for mypy. Here is my work on this so far: https://github.com/pypa/pip/compare/master...xmunoz:slots-attrs

Tests pass on the first two models that I converted, but mypy linting fails with nonsense.
```
src/pip/_internal/models/index.py:30: error: Too many arguments for "PackageIndex"
src/pip/_internal/models/index.py:30: error: Unexpected keyword argument "file_storage_domain" for "PackageIndex"
src/pip/_internal/models/index.py:33: error: Too many arguments for "PackageIndex"
src/pip/_internal/models/index.py:33: error: Unexpected keyword argument "file_storage_domain" for "PackageIndex"
src/pip/_internal/locations.py:188: error: Unexpected keyword argument "platlib" for "Scheme"
src/pip/_internal/locations.py:188: error: Unexpected keyword argument "purelib" for "Scheme"
src/pip/_internal/locations.py:188: error: Unexpected keyword argument "headers" for "Scheme"
src/pip/_internal/locations.py:188: error: Unexpected keyword argument "scripts" for "Scheme"
src/pip/_internal/locations.py:188: error: Unexpected keyword argument "data" for "Scheme"
Found 9 errors in 2 files (checked 112 source files)
```
A brief peruse at [mypy's issues](https://github.com/python/mypy/issues) reveal incomplete support for `attrs`.

--- a/src/pip/_internal/models/candidate.py
+++ b/src/pip/_internal/models/candidate.py
@@ -12,6 +12,8 @@ class InstallationCandidate(KeyBasedCompareMixin):
     """Represents a potential "candidate" for installation.
     """
 
+    __slots__ = ["name", "version", "link"]
+
     def __init__(self, name, version, link):
         # type: (str, str, Link) -> None
         self.name = name

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -26,7 +26,7 @@ class FormatControl(object):
     def __eq__(self, other):
         # type: (object) -> bool
         if not isinstance(other, self.__class__):
-            return False
+            return NotImplemented
 
         if self.__slots__ != other.__slots__:
             return False

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -3,14 +3,13 @@ import operator
 from pip._vendor.packaging.utils import canonicalize_name
 
 from pip._internal.exceptions import CommandError
-from pip._internal.utils.models import Base
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
     from typing import Optional, Set, FrozenSet
 
 
-class FormatControl(Base):
+class FormatControl(object):
     """Helper for managing formats from which a package can be installed.
     """
 

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -1,15 +1,19 @@
+import operator
 from pip._vendor.packaging.utils import canonicalize_name
 
 from pip._internal.exceptions import CommandError
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+from pip._internal.utils.models import Base
 
 if MYPY_CHECK_RUNNING:
     from typing import Optional, Set, FrozenSet
 
 
-class FormatControl(object):
+class FormatControl(Base):
     """Helper for managing formats from which a package can be installed.
     """
+
+    __slots__ = ["no_binary", "only_binary"]
 
     def __init__(self, no_binary=None, only_binary=None):
         # type: (Optional[Set[str]], Optional[Set[str]]) -> None
@@ -23,7 +27,12 @@ class FormatControl(object):
 
     def __eq__(self, other):
         # type: (object) -> bool
-        return self.__dict__ == other.__dict__
+        if isinstance(other, self.__class__):
+            if self.__slots__ == other.__slots__:
+                 attr_getters = [operator.attrgetter(attr) for attr in self.__slots__]
+                 return all(getter(self) == getter(other) for getter in attr_getters)
+
+        return False
 
     def __ne__(self, other):
         # type: (object) -> bool

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -1,4 +1,5 @@
 import operator
+
 from pip._vendor.packaging.utils import canonicalize_name
 
 from pip._internal.exceptions import CommandError

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -26,10 +26,10 @@ class FormatControl(object):
     def __eq__(self, other):
         # type: (object) -> bool
         if not isinstance(other, self.__class__):
-            return NotImplemented
+            return False
 
         if self.__slots__ != other.__slots__:
-            return NotImplemented
+            return False
 
         return all(
             getattr(self, k) == getattr(other, k)

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -1,5 +1,3 @@
-import operator
-
 from pip._vendor.packaging.utils import canonicalize_name
 
 from pip._internal.exceptions import CommandError

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -29,7 +29,7 @@ class FormatControl(object):
         # type: (object) -> bool
         if not isinstance(other, self.__class__):
             return NotImplemented
-            
+
         if self.__slots__ != other.__slots__:
             return NotImplemented
 

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -27,14 +27,16 @@ class FormatControl(object):
 
     def __eq__(self, other):
         # type: (object) -> bool
-        if isinstance(other, self.__class__):
-            if self.__slots__ == other.__slots__:
-                attr_getters = [operator.attrgetter(attr)
-                                for attr in self.__slots__]
-                return all(getter(self) == getter(other)
-                           for getter in attr_getters)
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+            
+        if self.__slots__ != other.__slots__:
+            return NotImplemented
 
-        return False
+        return all(
+            getattr(self, k) == getattr(other, k)
+            for k in self.__slots__
+        )
 
     def __ne__(self, other):
         # type: (object) -> bool

--- a/src/pip/_internal/models/format_control.py
+++ b/src/pip/_internal/models/format_control.py
@@ -2,8 +2,8 @@ import operator
 from pip._vendor.packaging.utils import canonicalize_name
 
 from pip._internal.exceptions import CommandError
-from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 from pip._internal.utils.models import Base
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
     from typing import Optional, Set, FrozenSet
@@ -29,8 +29,10 @@ class FormatControl(Base):
         # type: (object) -> bool
         if isinstance(other, self.__class__):
             if self.__slots__ == other.__slots__:
-                 attr_getters = [operator.attrgetter(attr) for attr in self.__slots__]
-                 return all(getter(self) == getter(other) for getter in attr_getters)
+                attr_getters = [operator.attrgetter(attr)
+                                for attr in self.__slots__]
+                return all(getter(self) == getter(other)
+                           for getter in attr_getters)
 
         return False
 

--- a/src/pip/_internal/models/index.py
+++ b/src/pip/_internal/models/index.py
@@ -1,9 +1,7 @@
 from pip._vendor.six.moves.urllib import parse as urllib_parse
 
-from pip._internal.utils.models import Base
 
-
-class PackageIndex(Base):
+class PackageIndex(object):
     """Represents a Package Index and provides easier access to endpoints
     """
 

--- a/src/pip/_internal/models/index.py
+++ b/src/pip/_internal/models/index.py
@@ -1,4 +1,5 @@
 from pip._vendor.six.moves.urllib import parse as urllib_parse
+
 from pip._internal.utils.models import Base
 
 
@@ -6,7 +7,8 @@ class PackageIndex(Base):
     """Represents a Package Index and provides easier access to endpoints
     """
 
-    __slots__ = ["url", "netloc", "simple_url", "pypi_url", "file_storage_domain"]
+    __slots__ = ['url', 'netloc', 'simple_url', 'pypi_url',
+                 'file_storage_domain']
 
     def __init__(self, url, file_storage_domain):
         # type: (str, str) -> None

--- a/src/pip/_internal/models/index.py
+++ b/src/pip/_internal/models/index.py
@@ -1,9 +1,12 @@
 from pip._vendor.six.moves.urllib import parse as urllib_parse
+from pip._internal.utils.models import Base
 
 
-class PackageIndex(object):
+class PackageIndex(Base):
     """Represents a Package Index and provides easier access to endpoints
     """
+
+    __slots__ = ["url", "netloc", "simple_url", "pypi_url", "file_storage_domain"]
 
     def __init__(self, url, file_storage_domain):
         # type: (str, str) -> None

--- a/src/pip/_internal/models/link.py
+++ b/src/pip/_internal/models/link.py
@@ -24,6 +24,14 @@ class Link(KeyBasedCompareMixin):
     """Represents a parsed link from a Package Index's simple URL
     """
 
+    __slots__ = [
+        "_parsed_url",
+        "_url",
+        "comes_from",
+        "requires_python",
+        "yanked_reason",
+    ]
+
     def __init__(
         self,
         url,                   # type: str

--- a/src/pip/_internal/models/link.py
+++ b/src/pip/_internal/models/link.py
@@ -30,6 +30,7 @@ class Link(KeyBasedCompareMixin):
         "comes_from",
         "requires_python",
         "yanked_reason",
+        "cache_link_parsing",
     ]
 
     def __init__(

--- a/src/pip/_internal/models/scheme.py
+++ b/src/pip/_internal/models/scheme.py
@@ -5,11 +5,15 @@ For a general overview of available schemes and their context, see
 https://docs.python.org/3/install/index.html#alternate-installation.
 """
 
+from pip._internal.utils.models import Base
 
-class Scheme(object):
+class Scheme(Base):
     """A Scheme holds paths which are used as the base directories for
     artifacts associated with a Python package.
     """
+
+    __slots__ = ['platlib', 'purelib', 'headers', 'scripts', 'data']
+
     def __init__(
         self,
         platlib,  # type: str

--- a/src/pip/_internal/models/scheme.py
+++ b/src/pip/_internal/models/scheme.py
@@ -7,6 +7,7 @@ https://docs.python.org/3/install/index.html#alternate-installation.
 
 from pip._internal.utils.models import Base
 
+
 class Scheme(Base):
     """A Scheme holds paths which are used as the base directories for
     artifacts associated with a Python package.

--- a/src/pip/_internal/models/scheme.py
+++ b/src/pip/_internal/models/scheme.py
@@ -5,10 +5,8 @@ For a general overview of available schemes and their context, see
 https://docs.python.org/3/install/index.html#alternate-installation.
 """
 
-from pip._internal.utils.models import Base
 
-
-class Scheme(Base):
+class Scheme(object):
     """A Scheme holds paths which are used as the base directories for
     artifacts associated with a Python package.
     """

--- a/src/pip/_internal/models/search_scope.py
+++ b/src/pip/_internal/models/search_scope.py
@@ -9,6 +9,7 @@ from pip._vendor.six.moves.urllib import parse as urllib_parse
 from pip._internal.models.index import PyPI
 from pip._internal.utils.compat import has_tls
 from pip._internal.utils.misc import normalize_path, redact_auth_from_url
+from pip._internal.utils.models import Base
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
@@ -18,11 +19,13 @@ if MYPY_CHECK_RUNNING:
 logger = logging.getLogger(__name__)
 
 
-class SearchScope(object):
+class SearchScope(Base):
 
     """
     Encapsulates the locations that pip is configured to search.
     """
+
+    __slots__ = ["find_links", "index_urls"]
 
     @classmethod
     def create(

--- a/src/pip/_internal/models/search_scope.py
+++ b/src/pip/_internal/models/search_scope.py
@@ -9,7 +9,6 @@ from pip._vendor.six.moves.urllib import parse as urllib_parse
 from pip._internal.models.index import PyPI
 from pip._internal.utils.compat import has_tls
 from pip._internal.utils.misc import normalize_path, redact_auth_from_url
-from pip._internal.utils.models import Base
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
@@ -19,7 +18,7 @@ if MYPY_CHECK_RUNNING:
 logger = logging.getLogger(__name__)
 
 
-class SearchScope(Base):
+class SearchScope(object):
 
     """
     Encapsulates the locations that pip is configured to search.

--- a/src/pip/_internal/models/selection_prefs.py
+++ b/src/pip/_internal/models/selection_prefs.py
@@ -1,3 +1,4 @@
+from pip._internal.utils.models import Base
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
@@ -5,12 +6,13 @@ if MYPY_CHECK_RUNNING:
     from pip._internal.models.format_control import FormatControl
 
 
-class SelectionPreferences(object):
-
+class SelectionPreferences(Base):
     """
     Encapsulates the candidate selection preferences for downloading
     and installing files.
     """
+
+    __slots__ = ['allow_yanked', 'allow_all_prereleases', 'format_control', 'prefer_binary', 'ignore_requires_python']
 
     # Don't include an allow_yanked default value to make sure each call
     # site considers whether yanked releases are allowed. This also causes

--- a/src/pip/_internal/models/selection_prefs.py
+++ b/src/pip/_internal/models/selection_prefs.py
@@ -1,4 +1,3 @@
-from pip._internal.utils.models import Base
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
@@ -6,7 +5,7 @@ if MYPY_CHECK_RUNNING:
     from pip._internal.models.format_control import FormatControl
 
 
-class SelectionPreferences(Base):
+class SelectionPreferences(object):
     """
     Encapsulates the candidate selection preferences for downloading
     and installing files.

--- a/src/pip/_internal/models/selection_prefs.py
+++ b/src/pip/_internal/models/selection_prefs.py
@@ -12,7 +12,8 @@ class SelectionPreferences(Base):
     and installing files.
     """
 
-    __slots__ = ['allow_yanked', 'allow_all_prereleases', 'format_control', 'prefer_binary', 'ignore_requires_python']
+    __slots__ = ['allow_yanked', 'allow_all_prereleases', 'format_control',
+                 'prefer_binary', 'ignore_requires_python']
 
     # Don't include an allow_yanked default value to make sure each call
     # site considers whether yanked releases are allowed. This also causes

--- a/src/pip/_internal/models/target_python.py
+++ b/src/pip/_internal/models/target_python.py
@@ -5,7 +5,6 @@ from pip._internal.utils.compatibility_tags import (
     version_info_to_nodot,
 )
 from pip._internal.utils.misc import normalize_version_info
-from pip._internal.utils.models import Base
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
@@ -14,7 +13,7 @@ if MYPY_CHECK_RUNNING:
     from pip._vendor.packaging.tags import Tag
 
 
-class TargetPython(Base):
+class TargetPython(object):
 
     """
     Encapsulates the properties of a Python interpreter one is targeting

--- a/src/pip/_internal/models/target_python.py
+++ b/src/pip/_internal/models/target_python.py
@@ -5,6 +5,7 @@ from pip._internal.utils.compatibility_tags import (
     version_info_to_nodot,
 )
 from pip._internal.utils.misc import normalize_version_info
+from pip._internal.utils.models import Base
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
@@ -13,12 +14,22 @@ if MYPY_CHECK_RUNNING:
     from pip._vendor.packaging.tags import Tag
 
 
-class TargetPython(object):
+class TargetPython(Base):
 
     """
     Encapsulates the properties of a Python interpreter one is targeting
     for a package install, download, etc.
     """
+
+    __slots__ = [
+        "_given_py_version_info",
+        "abi",
+        "implementation",
+        "platform",
+        "py_version",
+        "py_version_info",
+        "_valid_tags",
+    ]
 
     def __init__(
         self,

--- a/src/pip/_internal/utils/models.py
+++ b/src/pip/_internal/utils/models.py
@@ -7,15 +7,8 @@ import operator
 
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
-if MYPY_CHECK_RUNNING:
-    from typing import List
 
-
-class Base(object):
-    __slots__ = []  # type: List[str]
-
-
-class KeyBasedCompareMixin(Base):
+class KeyBasedCompareMixin(object):
     """Provides comparison capabilities that is based on a key
     """
 

--- a/src/pip/_internal/utils/models.py
+++ b/src/pip/_internal/utils/models.py
@@ -5,8 +5,6 @@
 
 import operator
 
-from pip._internal.utils.typing import MYPY_CHECK_RUNNING
-
 
 class KeyBasedCompareMixin(object):
     """Provides comparison capabilities that is based on a key

--- a/src/pip/_internal/utils/models.py
+++ b/src/pip/_internal/utils/models.py
@@ -4,7 +4,11 @@
 # mypy: disallow-untyped-defs=False
 
 import operator
-from typing import List
+
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+if MYPY_CHECK_RUNNING:
+    from typing import List
 
 
 class Base(object):

--- a/src/pip/_internal/utils/models.py
+++ b/src/pip/_internal/utils/models.py
@@ -4,10 +4,12 @@
 # mypy: disallow-untyped-defs=False
 
 import operator
+from typing import List
 
 
 class Base(object):
-    __slots__ = []
+    __slots__ = []  # type: List[str]
+
 
 class KeyBasedCompareMixin(Base):
     """Provides comparison capabilities that is based on a key

--- a/src/pip/_internal/utils/models.py
+++ b/src/pip/_internal/utils/models.py
@@ -6,9 +6,13 @@
 import operator
 
 
-class KeyBasedCompareMixin(object):
+class Base(object):
+    __slots__ = []
+
+class KeyBasedCompareMixin(Base):
     """Provides comparison capabilities that is based on a key
     """
+    __slots__ = ['_compare_key', '_defining_class']
 
     def __init__(self, key, defining_class):
         self._compare_key = key

